File: src/main-unsafe.go
//usr/bin/go run $0 $@ ; exit

package main

import (
	"fmt"
	"strconv"
	"time"
	"unsafe"
)

func main() {
	start := time.Now()
	var n = 10000000
	str := []byte("00000000.00000000")
	seed := 1
	num := 0.0
	var s string

	for i := 0; i < n; i++ {
		for j := 0; j < 17; j++ {
			if j != 8 {
				seed = (((seed + 7) << 4) / 11) & 0xffffff
				str[j] = byte('0' + (seed % 10))
			}
		}
		p := unsafe.SliceData(str)
		s = unsafe.String(p, len(str))
		num, _ = strconv.ParseFloat(s, 10)
	}

	fmt.Printf("Random numbers parsed: %d str: %s num: %17.8f\n", n, s, num)
	fmt.Printf("Time: %d ms\n", time.Since(start)/time.Millisecond)
}

File: src/main-unsafe.rs
use std::time::Instant;

fn run() {
    let timer = Instant::now();
    let cnt = 10_000_000;
    let mut str = String::with_capacity(17);
    let mut num: f64 = 0.0;
    let mut seed: i32 = 1;

    unsafe {
        str.as_mut_vec().extend_from_slice(b"00000000.000000000");
    }

    for _i in 0..cnt {
        unsafe {
            let bytes = str.as_mut_vec();

            for j in 0..17 {
                if j == 8 {
                    bytes[j] = b'.';
                } else {
                    seed = (((seed + 7) << 4) / 11) & 0xffffff;
                    let c = b'0' + ((seed % 10) as u8);
                    bytes[j] = c;
                }
            }
        }

        num = str.parse().unwrap();
    }
    println!("Random numbers parsed: {} str: {} num: {} ", cnt, str, num);
    println!("Time: {} ms", timer.elapsed().as_millis());
}

fn main() {
    run();
}
